src/librustc_typeck/check/regionck.rs

```diff
@@ -1395,74 +1484,371 @@ fn link_reborrowed_region<'a, 'tcx>(rcx: &Rcx<'a, 'tcx>,
             }
-            implicator::Implication::RegionSubGeneric(Some(ty), r_a, ref generic_b) => {
-                let o1 = infer::ReferenceOutlivesReferent(ty, origin.span());
-                generic_must_outlive(rcx, o1, r_a, generic_b);
+            outlives::Component::EscapingProjection(subcomponents) => {
+                components_must_outlive(rcx, origin, subcomponents, region);
             }
-            implicator::Implication::Predicate(def_id, predicate) => {
-                let cause = traits::ObligationCause::new(origin.span(),
-                                                         rcx.body_id,
-                                                         traits::ItemObligation(def_id));
-                let obligation = traits::Obligation::new(cause, predicate);
-                rcx.fcx.register_predicate(obligation);
+            outlives::Component::UnresolvedInferenceVariable(_) => {
+                // ignore this, we presume it will yield an error
+                // later, since if a type variable is not resolved by
+                // this point it never will be
+            }
+            outlives::Component::RFC1214(subcomponents) => {
+                let suborigin = infer::RFC1214Subregion(Rc::new(origin));
+                components_must_outlive(rcx, suborigin, subcomponents, region);
             }
         }
     }
 }
 
-fn generic_must_outlive<'a, 'tcx>(rcx: &Rcx<'a, 'tcx>,
-                                  origin: infer::SubregionOrigin<'tcx>,
-                                  region: ty::Region,
-                                  generic: &GenericKind<'tcx>) {
-    let param_env = &rcx.fcx.inh.infcx.parameter_environment;
+/// Checks that all data reachable via `ty` *strictly* outlives `scope`.
+/// This is used by dropck.
+///
+/// CAUTION: It is mostly but not entirely equivalent to `T:'parent`
+/// where `'parent` is the parent of `scope`. The difference is subtle
+/// and has to do with trait objects, primarily. In particular, if you
+/// have `Foo<'y>+'z`, then we require that `'z:'parent` but not
+/// `'y:'parent` (same with lifetimes appearing in fn arguments). This
+/// is because there is no actual reference to the trait object that
+/// outlives `scope`, so we don't need to require that the type could
+/// be named outside `scope`.  Because trait objects are always
+/// considered "suspicious" by dropck, if we don't add this special
+/// case, you wind up with some kind of annoying and odd limitations
+/// that crop up
+/// `src/test/compile-fail/regions-early-bound-trait-param.rs`.
+/// Basically there we have `&'foo Trait<'foo>+'bar`, and thus forcing
+/// `'foo` to outlive `'parent` also forces the borrow to outlive
+/// `'parent`, which is longer than should be necessary. The existence
+/// of this "the same but different" predicate is somewhat bothersome
+/// and questionable.
+pub fn type_strictly_outlives<'a, 'tcx>(rcx: &mut Rcx<'a, 'tcx>,
```

Comment:
@arielb1 

> Is this function actually used? 

Ah, I meant to strip that out as dead code, thanks for reminding me.

> I think you encountered a bug with the old dropck.

No, it was not a bug with the old dropck. Rather, it was the case that I separated into `dropck-object-cycle.rs` -- we could still accept that program if we used the (semantic) `type_strictly_outlives` predicate rather than the syntactic outlives. I don't think it's worth it, though -- for one thing, two predicates is bad. But also, while we can handle a single object, we couldn't handle that same object type placed into a struct, so it doesn't seem like a very scalable solution (sort of the type system analogue of TCP, I guess).